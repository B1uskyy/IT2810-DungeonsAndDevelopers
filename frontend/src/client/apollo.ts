import { ApolloClient, createHttpLink, InMemoryCache } from '@apollo/client';
import { setContext } from '@apollo/client/link/context';

const httpLink = createHttpLink({
  uri: 'http://localhost:4000/graphql',
});

const authLink = setContext((_, { headers }) => {
  const token = localStorage.getItem('token');
  return {
    headers: {
      ...headers,
      authorization: token ? `Bearer ${token}` : '',
    },
  };
});

const client = new ApolloClient({
  link: authLink.concat(httpLink),
  devtools: {
    enabled: true,
  },
  cache: new InMemoryCache({
    typePolicies: {
      Query: {
        fields: {
          user: {
            keyArgs: ['id'],
            merge(existing = {}, incoming) {
              return { ...existing, ...incoming };
            },
          },
        },
      },
      User: {
        keyFields: ['id'],
        fields: {
          favoritedMonsters: {
            merge: false,
          },
          dungeonName: {
            merge: false,
          },
          class: {
            merge(_, incoming) {
              return incoming;
            },
          },

          race: {
            merge(_, incoming) {
              return incoming;
            },
          },
        },
      },
      Class: {
        keyFields: ['name'],
      },
      Race: {
        keyFields: ['name'],
      },
      Monster: {
        keyFields: ['id'],
      },
    },
  }),
});

export default client;
