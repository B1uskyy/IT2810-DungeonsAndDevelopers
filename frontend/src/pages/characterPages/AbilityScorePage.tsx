import AbilityScoreCard from '../../components/CharacterCustomization/AbilityScoreCard.tsx';
import CharacterButton from '../../components/CharacterCustomization/CharacterButton.tsx';
import Navbar from '../../components/Navbar/Navbar.tsx';


export default function AbilityScorePage(){
  return (
    <>
      <Navbar />
      <section className="flex items-center justify-end h-screen z-0 before:absolute before:inset-0 before:bg-bluedrake before:bg-cover before:bg-center before:animate-background-zoom  before:z-0">
          <CharacterButton></CharacterButton>
      </section>
      <section className="flex flex-col items-center w-full bg-gradient-to-b from-dragongreen to-black">
        <AbilityScoreCard></AbilityScoreCard>
      </section>

    </>
  );
}