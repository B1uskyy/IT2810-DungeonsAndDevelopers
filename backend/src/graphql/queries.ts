import { gql } from '@apollo/client';

export const GET_MONSTERS = gql`
    query GetMonsters($searchTerm: String, $offset: Int, $limit: Int, $types: [String!]) {
        monsters(searchTerm: $searchTerm, offset: $offset, limit: $limit, types: $types) {
            monsters {
                id
                name
                size
                type
                alignment
                hit_points
                image
            }
            totalMonsters
        }
    }
`;

export const GET_USER_DUNGEON = gql`
  query GetUserDungeon($userId: ID!) {
    user(id: $userId) {
      dungeonName
      favoritedMonsters {
        id
      }
    }
  }
`;

export const UPDATE_DUNGEON_NAME = gql`
  mutation UpdateDungeonName($userId: ID!, $dungeonName: String!) {
    updateDungeonName(userId: $userId, dungeonName: $dungeonName) {
      dungeonName
    }
  }
`;

export const GET_USER_FAVORITES = gql`
  query GetUserFavorites($userId: ID!) {
    user(id: $userId) {
      favoritedMonsters {
        id
        name
        size
        type
        alignment
        hit_points
        image
      }
    }
  }
`;

export const ADD_FAVORITE_MONSTER = gql`
  mutation AddFavoriteMonster($userId: ID!, $monsterId: ID!) {
    addFavoriteMonster(userId: $userId, monsterId: $monsterId) {
      favoritedMonsters {
        id
      }
    }
  }
`;

export const REMOVE_FAVORITE_MONSTER = gql`
  mutation RemoveFavoriteMonster($userId: ID!, $monsterId: ID!) {
    removeFavoriteMonster(userId: $userId, monsterId: $monsterId) {
      favoritedMonsters {
        id
      }
    }
  }
`;

export const GET_MONSTER_REVIEWS = gql`
  query GetMonsterReviews($monsterId: ID!) {
    monster(id: $monsterId) {
      reviews {
        id
        user {
          id
          userName
        }
        difficulty
        description
        createdAt
      }
    }
  }
`;

export const ADD_REVIEW = gql`
  mutation AddReview($monsterId: ID!, $review: ReviewInput!) {
    addReview(monsterId: $monsterId, review: $review) {
      reviews {
        user {
          id
          userName
        }
        difficulty
        description
        createdAt
      }
    }
  }
`;

export const DELETE_REVIEW = gql`
  mutation DeleteReview($monsterId: ID!, $reviewId: ID!) {
    deleteReview(monsterId: $monsterId, reviewId: $reviewId) {
      reviews {
        id
        user {
          id
          userName
        }
        difficulty
        description
        createdAt
      }
    }
  }
`;

export const UPDATE_REVIEW = gql`
  mutation UpdateReview($monsterId: ID!, $reviewId: ID!, $review: ReviewInput!) {
    updateReview(monsterId: $monsterId, reviewId: $reviewId, review: $review) {
      id
      user {
        id
        userName
      }
      difficulty
      description
      createdAt
    }
  }
`;

export const CHECK_USERNAME = gql`
    query checkUsername($userName: String!) {
        checkUsername(userName: $userName)
    }
`;

export const GET_USER_EQUIPMENT = gql`
  query GetUserEquipment($userId: ID!) {
    user(id: $userId) {
      id
      equipments {
        id
        index
        name
        category
        value
      }
    }
  }
`;

export const ADD_EQUIPMENT_TO_CHARACTER = gql`
  mutation AddEquipmentToCharacter($userId: ID!, $equipmentId: ID!) {
    addEquipmentToCharacter(userId: $userId, equipmentId: $equipmentId) {
      id
      equipments {
        id
        index
        name
        category
        value
      }
    }
  }
`;

export const REMOVE_EQUIPMENT_FROM_CHARACTER = gql`
  mutation RemoveEquipmentFromCharacter($userId: ID!, $equipmentId: ID!) {
    removeEquipmentFromCharacter(userId: $userId, equipmentId: $equipmentId) {
      id
      equipments {
        id
        index
        name
        category
        value
      }
    }
  }
`;

export const CREATE_USER = gql`
  mutation createUser($userName: String!) {
    createUser(userName: $userName) {
      token
      user {
        id
        userName
        class {
          name
        }
        race {
          name
        }
        abilityScores {
          name
          score
        }
        equipments {
          name
        }
      }
    }
  }
`;

export const LOGIN_USER = gql`
  mutation loginUser($userName: String!) {
    loginUser(userName: $userName) {
      token
      user {
        id
        userName
        class {
          name
        }
        race {
          name
        }
        abilityScores {
          name
          score
        }
        equipments {
          name
        }
      }
    }
  }
`;
export const UPDATE_USER_CLASS = gql`
    mutation UpdateUserClass($userId: ID!, $classId:ID!){
        updateUserClass(userId:$userId, classId: $classId){
            id
            class{
                name
                index
                hit_die
            }
        }
    }
`;

export const GET_CLASSES = gql`
    query GetClasses($offset: Int, $limit: Int) {
        classes(offset: $offset, limit: $limit) {
            classes {
                id
                index
                name
                hit_die
                skills
            }
            totalClasses
        }
    }
`;

export const GET_USER_CLASS = gql`
    query GetUserClass($userId: ID!) {
        user(id: $userId) {
            class {
                id
                name
                index
                hit_die
            }
        }
    }
`;

export const UPDATE_USER_RACE = gql`
    mutation UpdateUserRace($userId: ID!, $raceId: ID!) {
        updateUserRace(userId: $userId, raceId: $raceId) {
            id
            race {
                name
                speed
                alignment
                size
                img
            }
        }
    }
`;

export const GET_RACES = gql`
    query GetRaces($offset: Int, $limit: Int) {
        races(offset: $offset, limit: $limit) {
            races {
                id
                index
                name
                speed
                alignment
                size
                img
            }
            totalRaces
        }
    }
`;

export const GET_USER_RACE = gql`
    query GetUserRace($userId: ID!) {
        user(id: $userId) {
            id
            race {
                id
                name
                speed
                alignment
                size
                img
            }
        }
    }
`;