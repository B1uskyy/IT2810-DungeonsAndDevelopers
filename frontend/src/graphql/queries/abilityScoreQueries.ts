import { gql } from '@apollo/client';

export const GET_ABILITYSCORES = gql`
  query GetAbilityScores($offset: Int, $limit: Int) {
    abilities(offset: $offset, limit: $limit) {
      abilities {
        index
        name
        skills
      }
      totalAbilities
    }
  }
`;
