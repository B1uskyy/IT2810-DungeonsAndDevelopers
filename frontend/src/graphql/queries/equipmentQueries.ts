import { gql } from '@apollo/client';

export const GET_EQUIPMENT_SUGGESTIONS = gql`
  query GetEquipmentSuggestions($searchTerm: String!) {
    equipments(searchTerm: $searchTerm, suggestionsOnly: true) {
      equipments {
        id
        name
      }
      totalCount
    }
  }
`;

export const GET_EQUIPMENTS = gql`
  query GetEquipments($searchTerm: String, $offset: Int, $limit: Int) {
    equipments(searchTerm: $searchTerm, offset: $offset, limit: $limit) {
      equipments {
        id
        index
        name
        category
        value
      }
      totalCount
    }
  }
`;
