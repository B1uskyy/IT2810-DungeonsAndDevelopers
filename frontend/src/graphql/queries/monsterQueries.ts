import { gql } from '@apollo/client';

export const GET_MONSTERS = gql`
  query GetMonsters(
    $searchTerm: String
    $offset: Int
    $limit: Int
    $types: [String!]
    $minHp: Int
    $maxHp: Int
    $sortOption: String
  ) {
    monsters(
      searchTerm: $searchTerm
      offset: $offset
      limit: $limit
      types: $types
      minHp: $minHp
      maxHp: $maxHp
      sortOption: $sortOption
    ) {
      monsters {
        id
        name
        size
        type
        alignment
        hit_points
        image
        averageDifficulty
        reviewsCount
      }
      totalMonsters
      minHp
      maxHp
    }
  }
`;

export const GET_MONSTER_TYPE_COUNTS = gql`
  query GetMonsterTypeCounts($minHp: Int, $maxHp: Int) {
    monsterTypeCounts(minHp: $minHp, maxHp: $maxHp) {
      type
      count
    }
  }
`;

export const GET_MONSTER_HP_RANGE = gql`
  query GetMonsterHpRange {
    monsterHpRange {
      minHp
      maxHp
    }
  }
`;

export const GET_MONSTER_SUGGESTIONS = gql`
  query GetMonsterSuggestions($searchTerm: String, $types: [String!], $minHp: Int, $maxHp: Int) {
    monsters(searchTerm: $searchTerm, types: $types, minHp: $minHp, maxHp: $maxHp) {
      monsters {
        id
        name
      }
    }
  }
`;

export const GET_MONSTER_REVIEWS = gql`
  query GetMonsterReviews($monsterId: ID!) {
    monster(id: $monsterId) {
      id
      reviews {
        id
        user {
          id
          userName
        }
        difficulty
        description
        createdAt
      }
    }
  }
`;
