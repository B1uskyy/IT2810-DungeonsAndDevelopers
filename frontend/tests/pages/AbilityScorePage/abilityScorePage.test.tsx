import { MockedProvider } from '@apollo/client/testing';
import { act, fireEvent, render, screen, waitFor } from '@testing-library/react';
import { AuthContext } from '../../../src/context/AuthContext.tsx';
import { GET_ABILITYSCORES, GET_ARRAY_SCORES, UPDATE_ABILITY_SCORES } from '../../../src/graphql/queries.ts';
import AbilityScorePage from '../../../src/pages/subPages/abilityScorePage.tsx';

import { BrowserRouter } from 'react-router-dom';
import { afterEach, vi } from 'vitest';

// Mock the useToast hook
const mockShowToast = vi.fn();
vi.mock('../../../src/hooks/useToast.ts', () => ({
  useToast: () => ({
    showToast: mockShowToast,
  }),
}));

// Mock data for GraphQL queries
const mocks = [
  {
    request: {
      query: GET_ARRAY_SCORES,
      variables: { userId: '1' },
    },
    result: {
      data: {
        getArrayScores: [10, 12, 14, 8, 15, 9],
      },
    },
  },
  {
    request: {
      query: UPDATE_ABILITY_SCORES,
      variables: { userId: '1', scores: [10, 12, 14, 8, 15, 9] },
    },
    result: {
      data: {
        updateAbilityScores: true,
      },
    },
  },
  {
    request: {
      query: GET_ABILITYSCORES, // Add GET_ABILITYSCORES here
      variables: { offset: 0, limit: 6 },
    },
    result: {
      data: {
        abilities: {
          abilities: [
            { index: 1, name: 'Strength', skills: ['Athletics'] },
            { index: 2, name: 'Dexterity', skills: ['Acrobatics', 'Stealth'] },
            { index: 3, name: 'Constitution', skills: [] },
            { index: 4, name: 'Intelligence', skills: ['Investigation'] },
            { index: 5, name: 'Wisdom', skills: ['Insight', 'Perception'] },
            { index: 6, name: 'Charisma', skills: ['Persuasion'] },
          ],
          totalAbilities: 6,
        },
      },
    },
  },
];

// Test Suite
describe('AbilityScorePage', () => {
  afterEach(() => {
    vi.restoreAllMocks();
  });
  const userId = '1';
  const token = 'mock-token'; // Mock token
  const userName = 'Mock User'; // Mock user name
  it('renders loading state initially', () => {
    render(
      <MockedProvider mocks={mocks} addTypename={false} defaultOptions={{ watchQuery: { fetchPolicy: 'no-cache' } }}>
        <AuthContext.Provider value={{ userId, token, userName, login: vi.fn(), logout: vi.fn() }}>
          <BrowserRouter>
            <AbilityScorePage />
          </BrowserRouter>
        </AuthContext.Provider>
      </MockedProvider>
    );
    expect(screen.getByText('Loading...')).toBeInTheDocument();
  });

  it('renders ability scores after loading', async () => {
    vi.useRealTimers();
    render(
      <MockedProvider mocks={mocks} addTypename={false}>
        <AuthContext.Provider value={{ userId, token, userName, login: vi.fn(), logout: vi.fn() }}>
          <BrowserRouter>
            <AbilityScorePage />
          </BrowserRouter>
        </AuthContext.Provider>
      </MockedProvider>
    );

    await waitFor(() => {
      expect(screen.getAllByRole('heading', { level: 2 }).length).toBe(6);
    });

    expect(screen.getAllByText('Skills required:')).toMatchSnapshot();
  });

  it('updates ability score on user interaction', async () => {
    vi.useRealTimers();
    render(
      <MockedProvider mocks={mocks} addTypename={false}>
        <AuthContext.Provider value={{ userId, token, userName, login: vi.fn(), logout: vi.fn() }}>
          <BrowserRouter>
            <AbilityScorePage />
          </BrowserRouter>
        </AuthContext.Provider>
      </MockedProvider>
    );

    await waitFor(() => {
      expect(screen.getAllByText('Skills required:')).toHaveLength(6);
    });
    await waitFor(() => {
      expect(screen.getAllByLabelText('ability score value')[0]).toHaveTextContent('10');
    });

    vi.useFakeTimers();

    const incrementButton = screen.getAllByLabelText('Increment')[0];
    fireEvent.mouseDown(incrementButton);
    act(() => {
      vi.advanceTimersByTime(300);
    });

    fireEvent.mouseUp(incrementButton);
    vi.useRealTimers();

    await waitFor(() => {
      expect(screen.getAllByLabelText('ability score value')[0]).toHaveTextContent('13');
    });
  });
});
