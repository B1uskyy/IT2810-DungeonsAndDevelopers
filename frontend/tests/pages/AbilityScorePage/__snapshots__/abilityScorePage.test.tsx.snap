// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AbilityScorePage > renders ability scores after loading 1`] = `
[
  <li
    class="bold text"
  >
    Skills required:
  </li>,
  <li
    class="bold text"
  >
    Skills required:
  </li>,
  <li
    class="bold text"
  >
    Skills required:
  </li>,
  <li
    class="bold text"
  >
    Skills required:
  </li>,
  <li
    class="bold text"
  >
    Skills required:
  </li>,
  <li
    class="bold text"
  >
    Skills required:
  </li>,
]
`;
